File: matrix_algorithms/src/matrices/Algoritmo13_IV5_EnhancedParallelBlock.java
package matrices;

import java.math.BigInteger;
import java.util.concurrent.*;

/**
 * Algoritmo 13: IV.5 Enhanced Parallel Block
 * Tipo: Iterativo por bloques con paralelismo mejorado - orden i-k-j
 * Descripcion: Mejora de IV.4 que asigna rangos de filas exclusivos
 * a cada hilo (sin solapamiento ni sincronizacion). Combina:
 *   - Orden i-k-j para reuso de registros
 *   - Particion de filas sin synchronized
 *   - Tamano de bloque adaptado al numero de hilos
 * Complejidad: O(n^3 / p) con minimo overhead de sincronizacion
 */
public class Algoritmo13_IV5_EnhancedParallelBlock {

    static final int BLOQUE = 64;
    static final int HILOS  = Runtime.getRuntime().availableProcessors();

    public static BigInteger[][] multiplicar(BigInteger[][] A, BigInteger[][] B) throws Exception {
        int n = A.length;
        BigInteger[][] C = new BigInteger[n][n];
        for (int i=0;i<n;i++) for (int j=0;j<n;j++) C[i][j]=BigInteger.ZERO;

        ExecutorService pool = Executors.newFixedThreadPool(HILOS);
        java.util.List<Future<?>> tareas = new java.util.ArrayList<>();

        int filasPorHilo = Math.max(1, n / HILOS);
        for (int h=0; h<HILOS; h++) {
            final int filaInicio = h * filasPorHilo;
            final int filaFin    = (h == HILOS-1) ? n : filaInicio + filasPorHilo;
            if (filaInicio >= n) break;
            tareas.add(pool.submit(() -> {
                for (int ii=filaInicio; ii<filaFin; ii+=BLOQUE)
                    for (int kk=0; kk<n; kk+=BLOQUE)
                        for (int jj=0; jj<n; jj+=BLOQUE) {
                            int iM=Math.min(ii+BLOQUE,filaFin);
                            int kM=Math.min(kk+BLOQUE,n), jM=Math.min(jj+BLOQUE,n);
                            for (int i=ii; i<iM; i++)
                                for (int k=kk; k<kM; k++) {
                                    BigInteger aik = A[i][k];
                                    for (int j=jj; j<jM; j++)
                                        C[i][j] = C[i][j].add(aik.multiply(B[k][j]));
                                }
                        }
            }));
        }
        pool.shutdown();
        for (Future<?> f : tareas) f.get();
        return C;
    }
}

File: matrix_algorithms_python/algoritmo13_IV5_enhanced_parallel_block.py
# Algoritmo 13: IV.5 Enhanced Parallel Block
# Complejidad: O(n^3 / p) con minimo overhead de sincronizacion
# Combina orden i-k-j + particion de filas sin solapamiento + sin locks

import threading
import os

BLOQUE = 64
HILOS  = os.cpu_count()

def multiplicar(A, B):
    n = len(A)
    C = [[0] * n for _ in range(n)]

    def procesar_rango(fila_inicio, fila_fin):
        for ii in range(fila_inicio, fila_fin, BLOQUE):
            for kk in range(0, n, BLOQUE):
                for jj in range(0, n, BLOQUE):
                    iM = min(ii + BLOQUE, fila_fin)
                    kM = min(kk + BLOQUE, n)
                    jM = min(jj + BLOQUE, n)
                    for i in range(ii, iM):
                        for k in range(kk, kM):
                            aik = A[i][k]  # reuso de registro
                            for j in range(jj, jM):
                                C[i][j] += aik * B[k][j]

    filas_por_hilo = max(1, n // HILOS)
    hilos = []
    for h in range(HILOS):
        inicio = h * filas_por_hilo
        fin    = n if h == HILOS - 1 else inicio + filas_por_hilo
        if inicio >= n:
            break
        t = threading.Thread(target=procesar_rango, args=(inicio, fin))
        hilos.append(t)
        t.start()
    for t in hilos:
        t.join()
    return C
